google/cloud/opentelemetry/resource_detector.cc

```diff
@@ -0,0 +1,241 @@
+}
+
+void CloudRun(opentelemetry::sdk::resource::ResourceAttributes& attributes,
+              nlohmann::json const& json) {
+  attributes.SetAttribute(sc::kCloudPlatform, "gcp_cloud_run");
+  SetEnvAttribute(attributes, sc::kFaasName, "K_SERVICE");
+  SetEnvAttribute(attributes, sc::kFaasVersion, "K_REVISION");
+  SetAttribute(attributes, sc::kFaasInstance, MD(json, {"instance", "id"}));
+}
+
+void Gae(opentelemetry::sdk::resource::ResourceAttributes& attributes,
+         nlohmann::json const& json) {
+  attributes.SetAttribute(sc::kCloudPlatform, "gcp_app_engine");
+  SetEnvAttribute(attributes, sc::kFaasName, "GAE_SERVICE");
+  SetEnvAttribute(attributes, sc::kFaasVersion, "GAE_VERSION");
+  SetEnvAttribute(attributes, sc::kFaasInstance, "GAE_INSTANCE");
+
+  auto zone = Tail(MD(json, {"instance", "zone"}));
+  SetAttribute(attributes, sc::kCloudAvailabilityZone, zone);
+  auto const pos = zone.rfind('-');
+  SetAttribute(attributes, sc::kCloudRegion, zone.substr(0, pos));
+}
+
+void Gce(opentelemetry::sdk::resource::ResourceAttributes& attributes,
+         nlohmann::json const& json) {
+  attributes.SetAttribute(sc::kCloudPlatform, "gcp_compute_engine");
+  SetAttribute(attributes, sc::kHostType,
+               Tail(MD(json, {"instance", "machineType"})));
+  SetAttribute(attributes, sc::kHostId, MD(json, {"instance", "id"}));
+  SetAttribute(attributes, sc::kHostName, MD(json, {"instance", "name"}));
+
+  auto zone = Tail(MD(json, {"instance", "zone"}));
+  attributes.SetAttribute(sc::kCloudAvailabilityZone, zone);
+  auto const pos = zone.rfind('-');
+  attributes.SetAttribute(sc::kCloudRegion, zone.substr(0, pos));
+}
+
+class ResourceDetector final
+    : public opentelemetry::sdk::resource::ResourceDetector {
+ public:
+  explicit ResourceDetector(otel_internal::HttpClientFactory factory,
+                            Options options)
+      : client_factory_(std::move(factory)), options_(std::move(options)) {}
+
+  opentelemetry::sdk::resource::Resource Detect() override {
+    auto client = client_factory_(options_);
+    rest_internal::RestRequest request;
+    request.SetPath(absl::StrCat(internal::GceMetadataScheme(), "://",
+                                 internal::GceMetadataHostname(),
+                                 "/computeMetadata/v1/"));
+    request.AddHeader("metadata-flavor", "Google");
+    request.AddQueryParameter("recursive", "true");
+    rest_internal::RestContext context;
+    auto response = client->Get(context, request);
+
+    auto log_error = [](Status const& status) {
+      GCP_LOG(INFO) << "Could not query the metadata server. status=" << status;
+      return opentelemetry::sdk::resource::Resource::GetEmpty();
+    };
+
+    if (!response) {
+      auto status = std::move(response).status();
+      if (status.code() == StatusCode::kUnavailable) {
+        // This is most likely a "curl: (6) Could not resolve host" error, which
```

Comment:
If only... GKE is notorious for returning kUnavailable.  You need a retry loop.

Reply:
I think you can assume that a successful parse is "good forever".  A failed attempt should be retried after a few minutes.

Reply:
> You need a retry loop.

Good call. The retry/backoff settings are not obvious to me.

> A failed attempt should be retried after a few minutes.

Hmmm... thinking out loud.... `Detect()` should be called once from within `ConfigureBasicTracing()` when we initialize the `TracerProvider`.

https://github.com/googleapis/google-cloud-cpp/blob/0c7f077b96f802c0e10f500a3e601b2663bcf353/google/cloud/opentelemetry/configure_basic_tracing.cc#L66

It seems wrong to block an application not running on GCP for a few minutes. And I think you are saying we can't distinguish between a transient error and the case where there is no MD server.

We could introduce some `GcpResourceDetectorOption` that determines whether we include the resource detector or not.

Another (bad) workaround would be to build the resource detector implementation into the trace exporter.

Reply:
When is `Detect()` called?  Or more important: how often?    If it is called only once, we need to retry right there on the spot.  If it is called each time something gets exported we can try to fetch the values at that time, with caching to avoid expensive operations.